### DISCUSSION TITLE
Default client policies for realms

### DISCUSSION THREAD

**Discussion by rmartinc** [keycloak/keycloak#27171] [Ideas]
Keycloak has been supporting [client policies](https://github.com/keycloak/keycloak-community/blob/main/design/client-policies.md) for a long time. We think that it is time to start adding some default policies to the realms (for example we need it for issue #26432). This thread is to try to do some brainstorming about the implementation. Initially on that issue I had thought about a very direct integration:

1. Create by default a client policy for protocol OIDC in each new realm (at creation time), for example `Openid-connect default` or similar. The same idea can be extended for SAML if necessary.
2. The policy contains conditions that only executes the policy if the client is OIDC (this allows us to differentiate by protocol) and if running in prod mode (dev-mode does not execute them and is not affected by the policies).
3. We add the executors we consider that are necessary in the policy for production environment (PKCE executor now, the executor can be improved at this point to make the integration easier).
4. The admins can do whatever they like (enable/disable/delete/modify) with these default policies. It's just a normal policy but created automatically with the realm.
5. We never add them to existing realms, just new realms. This allows us to add more policies/conditions/executors when needed and they only affect the new realms. Existing realms/installations continue working as before.

WDYT? Ideas?

**Comment by mposolda** [comment]
I can add that default client policies are important also if we want to finally get rid of deprecated "Client registration policies" . As until we have ability to have "default client policies" in the realm, we cannot use them to replace "default client registration policies", which are defined in the new realms.

Few more points:
- Can we have default client policies defined in the JSON file similarly like user-profile or client-profiles? So the policies from the JSON will be pre-created in new realms, but administrators will have possibility to later update/remove them. So it would work similarly like user-profile.

- IMO we can have same default configuration added for both "production" and "dev" profile. But maybe we can have client policy condition implementation, which will evaluate to true just for prod/dev profile? It can probably have flag whether to apply it for "prod" or "dev" profile. This will allow us to have client policy, which will be "effective" just for the production profile as the particular condition will evaluate to false for dev profile.

**Reply by rmartinc** [reply]
> Can we have default client policies defined in the JSON file similarly like user-profile or client-profiles?

The problem I see with this is that any future change in the default policies would be directly inherited by the existing realms that were using them. Affecting migrations. That's why I think it's better if we create the policies normally for the new realms. But the default policies can be defined in a json file and use that file to add them when creating the realm.

> IMO we can have same default configuration added for both "production" and "dev" profile. But maybe we can have client policy condition implementation, which will evaluate to true just for prod/dev profile?

Yes, exactly, that was my idea too. Having a condition which only executes the policies in prod-mode (we need to implement this new condition). So dev-mode is not affected by default and continue working like it's doing now.

**Reply by mposolda** [reply]
Yep, so we are aligned with having default client policies defined in some JSON. Now the question is how we apply them to the new realms and existing realms? I see we have options like:

1) Create default client policies from JSON always just when new realm is created and save it to the realm automatically. This means that existing realms won't be affected anyhow.

2) Use something like we have for user-profile. So when you create new realm, default client policies are not created in the realm. But they will be effectively used when realm does not have any client policies explicitly defined and already saved in the realm attribute. So the fact, that realm does not have any client policies in the DB means, that it should effectively use the default client policy configuration from the JSON. So once admin change something in the client policies, it will always save the client policy to the DB and from this point, default config from JSON is not used.
For this option, we will also not need to do anything regarding migration AFAIK? As for example:
- If someone already has client policy defined on the realm, the default from JSON won't be applied
- If someone does not yet use client policies, it will effectively apply the default from JSON configuration. So the behaviour may change for those guys as new default client policies will be effectively applied. Do you think this can be a problem?

**Reply by rmartinc** [reply]
> Do you think this can be a problem?

Yes, because if, for example, PKCE is enforced all the clients without it would not work after the migration. That's why I was trying to avoid this situation. In general it's the same for any security requirement.

But if we follow the idea of adding a new option with several security levels maybe it's OK adding the policies by default. In the end they can decided to start with no policies enabled in LOW or NONE level (or whatever name we decide to use). But why not using a realm property instead of a startup option? (Something like security level NONE, LAX, STRICT). Default to LAX or similar.

**Reply by mposolda** [reply]
Yes, true. Applying policies by default can affect behaviour of existing deployments after migration. So I think we can go with option (1) then?

**Comment by stianst** [comment]
We need to add a new built-in client policy profile for this, so guess the main question is what this should include. I would lean towards something that defined on `OAuth 2.1` (i.e. no password grant, PKCE required, no wildcards in redirect-uris).

Then I'd say we should apply this new profile to all clients by default, but only when running in production mode. For that I'd suggest we'd add something to `Profile` to make it easy to detect the mode. We may want to have multiple different modes like dev, lax, strict, or something like that. For example we could add:

`enum SecurityProfile { DEV, LAX, STRICT }`

Then make it available in `Profile#getSecurityProfile`, and configurable through something like `--security-level`, and have the `start-dev` use `--security-level=dev` and `start` use `--security-level=strict` by default, but allow folks to override it independent of the run command. @vmuzikar FIY

**Reply by stianst** [reply]
We may want the `security-profile` or whatever we call it to be configurable, say we have a set of built-in security profiles, but then you can change parts of it (or define your own). For example:

`--security-profile-https-required`
`--security-profile-enable-default-client-policy`

That might make it even more intuitive what the difference between prod and dev modes are, as currently it's hidden.

**Reply by mposolda** [reply]
+1 for this option. This will allow us to have client-policy-condition implementation, which can read the current security profile from the `SecurityProfile` enum. That condition can evaluate to true for specific security-profile, so will allow to effectively use some client policies just for the production mode (and other less strict policies just in the lax mode etc).

**Reply by vmuzikar** [reply]
@stianst My only concern here would be that `--security-level` might be a bit ambiguous name. I know it's intentional as we might want to use it for other stuff in the future too which makes sense, but on the other hand I'd like to avoid the same mistake we did e.g. with `--proxy` which was a "smart" option (defining a "proxy profile") but it became quickly too confusing what exactly does it set.

Would it perhaps make sense to start with a specific security profile for client policies (`--security-profile-client-policies`)? We could still in the future add some "main switch" `--security-profile` that'd set all other `--security-profile-*` to a defined value if they are not explicitly specified. This might bring some clarity and granularity for what "secure profile" affects.

**Reply by rmartinc** [reply]
Why not using a realm level attr instead of startup option? (NONE, LAX, STRICT) I think it's easier.

**Reply by vmuzikar** [reply]
@rmartinc It's a good point different realms might want different security profiles.

**Reply by mposolda** [reply]
@rmartinc @vmuzikar Sorry for late response. I consider security-profile more as a server-wide thing. So administrator can for instance test things in the `dev` security-profile when in development/testing environment, but in production, he wants to make sure that proper security is enabled by default for all realms. Looks to me similar to `start` VS `start-dev` options.

We can maybe have an option as well on realm level, which will allow to override server-wide switch, but not 100% sure if we need that. Maybe can be added later based on the demand?

**Reply by rmartinc** [reply]
I have been thinking about this and I think we have two options:

1)  The realm has a new attribute security-level or similar: DEV, LAX, STRICT.
* The policies are templated in a file and all (for all the levels) are always create to the realm.
* Upgrades modifies the template files but not existing realm policies.
* The policies are executed via condition to the domain attribute set.
* The admin can change the level to execute other set of policies and can also add/modify/delete the policies.
* Default value for the attribute is LAX/STRICT (whatever we decide).
* If no value no policy will be executed (migrated realms).

2) Use something to mark the initial security level at creation: DEV, LAX, STRICT.
* The policies are templated in a file and are always created to the realm, but only the ones for the desired security level.
* Upgrades modifies the template files but not existing realm policies.
* The policies does not have a condition on level because only the ones specified for the level are created.
* The admin can add/modify/delete the policies.
* Here we should think what we use to mark the level at creation time. It can be a `kc.sh` startup option as suggested or even a new parameter in the create option (something new that defaults to LAX/STRICT). I prefer to use something for each realm, so I vote for the latter.

Possibly in both solutions we need something to recreate policies from the console as the admins can modify them. Just in case they want to re-import the original (or updated by upgrades) policies. In general I prefer solution 2 because we don't need to create all level the policies in the realms so I will try to work on that option and submit a draft. There we can decide if it's OK.

**Reply by stianst** [reply]
I'd propose we call it a security profile rather than a security level. Initially we would only have built-in profiles `none`, `lax`, and `strict`.

Each security profile would consists of a number of policies, like `require https`, `enforce default client policies`, etc. In the future we may want to be able to allow people defining their own custom security profiles, where one could give it a name and the values for the policies (for example myprofile: enforce-default-client-policies=true, require-https=false).

As the aim here at least for now is to enforce globally for the server, this should be a server-wide thing that is configured through server options, and where `dev` and `prod` modes would use different security profiles by default.

It may make sense to be able to override the security profile at a realm level, but I see that as out of scope initially. It also brings complexities like at a realm level it should be possible to enforce a stricter, but not more relaxed profile than the server level default.

**Reply by mposolda** [reply]
@stianst @rmartinc If server is started with `strict` mode, do we want realm administrator to be able to update (or delete) the default client policies?

I understand it can be good if we don't allow it for security purposes. At the same time, I can imagine that realm administrator may want to do some small adjustements. For instance, if `strict` mode has something like `Always enforce PKCE for all clients`, but realm has some "legacy" client where the PKCE cannot be used. Then realm admin may want to update default client policy and remove `Always` condition and replace it with other condition to not include that legacy client. If realm admin cannot do that, the workaround would be to start server with mode `lax` or `none`, which is worse from the security point of view.

Because of this, I see the client policies rather similar to user-profile. So "default client policies" are automatically applied for the new realms (after realm creation) or for existing realm (during migration), but realm administrator may still want to update/delete default client policies. WDYT?

**Reply by stianst** [reply]
I'd say not as that would mean a server admin can not require PKCE for all clients. There may be a case for `default policies` vs `enforced policies` though.

**Comment by mposolda** [comment]
Created epic related to default client policies https://github.com/keycloak/keycloak/issues/27188 . We can add various tasks related to it.

**Comment by stianst** [comment]
I'm thinking something like:

in keycloak jars:

- strict-security-profile.json / name=strict
- strict-client-policies.json / name=strict

---

conf/security-profile.json
{
  name: my-custom
  tls-required: none
  client-policies: strict  
}

---

config/client-policies.json
{
  name: my-custom
}


user-profile-.json



start --security-profile=strict|mycustom|none
